pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_test.go

```diff
@@ -0,0 +1,417 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dnsrecord_test
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/extensions"
+	"github.com/gardener/gardener/pkg/logger"
+	mockclient "github.com/gardener/gardener/pkg/mock/controller-runtime/client"
+	mocktime "github.com/gardener/gardener/pkg/mock/go/time"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/extensions/dnsrecord"
+	gutil "github.com/gardener/gardener/pkg/utils/gardener"
+	"github.com/gardener/gardener/pkg/utils/test"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+const (
+	name                = "foo"
+	namespace           = "shoot--foo--bar"
+	secretName          = "testsecret"
+	extensionType       = "provider"
+	zone                = "zone"
+	dnsName             = "foo.bar.external.example.com"
+	address             = "1.2.3.4"
+	ttl           int64 = 300
+)
+
+var _ = Describe("DNSRecord", func() {
+	var (
+		ctrl *gomock.Controller
+
+		mockNow *mocktime.MockNow
+
+		c client.Client
+
+		values    *dnsrecord.Values
+		dnsRecord dnsrecord.Interface
+
+		dns    *extensionsv1alpha1.DNSRecord
+		secret *corev1.Secret
+
+		ctx     = context.TODO()
+		now     = time.Now()
+		log     = logger.NewNopLogger()
+		testErr = fmt.Errorf("test")
+
+		cleanup func()
+	)
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+
+		scheme := runtime.NewScheme()
+		Expect(extensionsv1alpha1.AddToScheme(scheme)).NotTo(HaveOccurred())
+		Expect(corev1.AddToScheme(scheme)).NotTo(HaveOccurred())
+		c = fake.NewClientBuilder().WithScheme(scheme).Build()
+
+		values = &dnsrecord.Values{
+			Name:       name,
+			Namespace:  namespace,
+			SecretName: secretName,
+			Type:       extensionType,
+			SecretData: map[string][]byte{
+				"foo": []byte("bar"),
+			},
+			Zone:       pointer.String(zone),
+			DNSName:    dnsName,
+			RecordType: extensionsv1alpha1.DNSRecordTypeA,
+			Values:     []string{address},
+			TTL:        pointer.Int64(ttl),
+		}
+		dnsRecord = dnsrecord.New(log, c, values, time.Millisecond, 250*time.Millisecond, 500*time.Millisecond)
```

Comment:
Nit: generally it's nicer to use a fake implementation of `retry.Ops` rather than relying on short intervals/timeouts (makes tests faster, more robust in CI environments). E.g. like:
https://github.com/gardener/gardener/blob/712a875b5965c9c87a8ce9c304c13733fe4b11ce/pkg/operation/botanist/component/extensions/dns/dnsentry_test.go#L72

Reply:
Done